ocaml/xapi-guard/lib/varstored_interface.ml

```diff
@@ -147,29 +143,69 @@ let () =
    * this is only needed for syscalls that would otherwise block *)
   Lwt_unix.set_pool_size 16
 
-let with_xapi f =
+let with_xapi ~cache f =
   Lwt_unix.with_timeout 120. (fun () -> SessionCache.with_session cache f)
 
 (* Unfortunately Cohttp doesn't provide us a way to know when it finished
  * creating the socket, and creating the socket is done asynchronously in an Lwt promise.
  * It only ever returns from server creation when the server is stopped.
  * Try actually connecting: the file could be present but nobody listening on the other side.
  * *)
-let rec wait_for_file_to_appear path =
-  Lwt.pause () >>= fun () ->
-  Lwt.try_bind
-    (fun () ->
-      Conduit_lwt_unix.connect ~ctx:Conduit_lwt_unix.default_ctx
-        (`Unix_domain_socket (`File path))
-    )
-    (fun (_, ic, oc) ->
+let rec wait_connectable path =
+  let* res =
+    Lwt_result.catch
+      (Conduit_lwt_unix.connect ~ctx:Conduit_lwt_unix.default_ctx
+         (`Unix_domain_socket (`File path))
+      )
+  in
+  match res with
+  | Ok (_, ic, oc) ->
       D.debug "Socket at %s works" path ;
-      Lwt_io.close oc >>= fun () -> Lwt_io.close ic
-    )
-    (fun e ->
-      D.debug "Waiting for file %s to appear (%s)" path (Printexc.to_string e) ;
-      Lwt_unix.sleep 0.1 >>= fun () -> wait_for_file_to_appear path
+      let* () = Lwt_io.close oc in
+      Lwt_io.close ic
+  | Error e ->
+      D.debug "Waiting for socket to be connectable at %s: %s" path
+        (Printexc.to_string e) ;
+      (* just in case Lwt_unix.listen doesn't get called soon enough,
+         avoid using up 100% CPU waiting for it,
+         better than Lwt.pause ()
+      *)
+      let* () = Lwt_unix.sleep 0.001 in
```

Comment:
The wait is a lot shorter now. Is this just for Lwt scheduling?

Reply:
This is for getting the unit tests to pass quickly while still yielding the CPU to other tasks, see https://github.com/xapi-project/xen-api/pull/4670#discussion_r844046467